nodejs/aws-infra/iam/index.ts

```diff
@@ -0,0 +1,195 @@
+        name: string,
+        id: pulumi.Input<string>,
+        args: UserArgs = {},
+        opts: pulumi.ComponentResourceOptions = {},
+    ) {
+        return new User(name, { ...args, user: aws.iam.User.get(name, id, {}, opts) }, opts);
+    }
+}
+
+type OverwriteUserArgs = utils.Overwrite<aws.iam.UserArgs, {
+    groupMembership?: UserGroupMembershipArgs;
+    user?: aws.iam.User;
+}>;
+
+/**
+ * The set of arguments for constructing a User resource.
+ */
+export interface UserArgs {
+    /**
+     * When destroying this user, destroy even if it
+     * has non-Terraform-managed IAM access keys, login profile or MFA devices. Without `force_destroy`
+     * a user with non-Terraform-managed access keys and login profile will fail to be destroyed.
+     */
+    readonly forceDestroy?: pulumi.Input<boolean>;
+
+    /**
+     * The IAM Groups this User belongs to.
+     */
+    groupMembership?: UserGroupMembershipArgs;
+
+    /**
+     * The user's name. The name must consist of upper and lowercase alphanumeric characters with no spaces. You can also include any of the following characters: `=,.@-_.`. User names are not distinguished by case. For example, you cannot create users named both "TESTUSER" and "testuser".
+     */
+    readonly name?: pulumi.Input<string>;
+
+    /**
+     * Path in which to create the user.
+     */
+    readonly path?: pulumi.Input<string>;
+
+    /**
+     * The ARN of the policy that is used to set the permissions boundary for the user.
+     */
+    readonly permissionsBoundary?: pulumi.Input<string>;
+
+    /**
+     * Key-value mapping of tags for the IAM user
+     */
+    readonly tags?: pulumi.Input<{
+        [key: string]: any;
+    }>;
+
+    /**
+     * An existing User to use for this awsinfra User.  If not provided, a default one will be
+     * created.
+     */
+    user?: aws.iam.User;
+}
+
+// Make sure our exported args shape is compatible with the overwrite shape we're trying to provide.
+const test1: string = utils.checkCompat<OverwriteUserArgs, UserArgs>();
+
+type OverwriteUserGroupMembershipArgs = utils.Overwrite<aws.iam.UserGroupMembershipArgs, {
+    user?: never;
+}>;
+
+/**
+ * The set of arguments for constructing a UserGroupMembership resource.
+ */
+export interface UserGroupMembershipArgs {
+    /**
+     * A list of [IAM Groups][1] to add the user to
+     */
+    readonly groups: pulumi.Input<pulumi.Input<string>[]>;
+}
+
+// Make sure our exported args shape is compatible with the overwrite shape we're trying to provide.
+const test2: string = utils.checkCompat<
+    OverwriteUserGroupMembershipArgs,
+    UserGroupMembershipArgs
+>();
+
+type OverwriteAccessKeyArgs = utils.Overwrite<aws.iam.AccessKeyArgs, { user?: never }>;
+
+/**
+ * The set of arguments for constructing a AccessKey resource.
+ */
+export interface AccessKeyArgs {
+    /**
+     * Either a base-64 encoded PGP public key, or a
+     * keybase username in the form `keybase:some_person_that_exists`.
```

Comment:
also, instead of this, should we consider having two exclusive properties, and just passing them along accordingly.  this is our abstraction, so we don't necessarliy have to overload concepts like this.